### PR TITLE
Add Numbers API snippets

### PR DESCRIPTION
**Pull request by marklewin**
Create Numbers API snippets:

- List owned numbers
- Search for available numbers
- Buy a number
- Update a number
- Cancel a number

### CODE REVIEW

**Comment by marklewin** [comment]
@benhayehudi I've used the same approach as the other Ruby code snippets. But I'd  be very grateful if you have any suggestions as to how these could be improved - especially with regard to working with responses, error handling, etc.

**Comment by marklewin** [comment]
Adding @mheap as reviewer in Ben's absence.

File: numbers/buy.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+NEXMO_NUMBER = ENV['NEXMO_NUMBER']
+
+require 'nexmo'
+
+client = Nexmo::Client.new(
+  api_key: NEXMO_API_KEY,
+  api_secret: NEXMO_API_SECRET
+)
+
+response = client.numbers.buy(
```
Comment (mheap) [inline numbers/buy.rb]: Is there anything in the response that we could output?

File: numbers/list.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+require 'nexmo'
+
+client = Nexmo::Client.new(
+  api_key: NEXMO_API_KEY,
+  api_secret: NEXMO_API_SECRET
+)
+
+response = client.numbers.list()
+
+puts response.inspect
```
Comment (mheap) [inline numbers/list.rb]: Using the response should be more detailed imo e.g.

```
response.each do |number|
  puts number['msisdn']
```

Not that *this is pseudocode* and I haven't actually checked what the response looks like

**Comment by timcraft** [comment]
@marklewin @mheap The `response.is_a? Nexmo::ClientError` lines won't work—Nexmo::ClientError is a generic error which will be raised as an exception. For buy/cancel/update calls the response has a couple of status fields, so the response handling should be something like this:

```ruby
response = client.numbers.buy(country: COUNTRY_CODE, msisdn: NEXMO_NUMBER)

if response.error_code == 200
  puts "Number purchased"
else
  puts "Error #{response.error_code} (#{response.error_code_label})"
end
```

There's nothing usable in the response for success responses. If the `response.error_code` field is *always* going to be 200 for a HTTP 200 response then it's redundant code which would be better left out. Otherwise a list of possible error codes in the API Reference would be really helpful.

**Comment by mheap** [comment]
@timcraft That's what I expected to happen, but testing in `irb` the exception is returned rather than raised.

**Comment by timcraft** [comment]
> That's what I expected to happen, but testing in `irb` the exception is returned rather than raised.

You're correct. And that's a bug, I'll get it fixed asap.

**Comment by mheap** [comment]
Fixing this will be a BC break, so maybe worth waiting until the
applications v2 release

**Comment by timcraft** [comment]
> Fixing this will be a BC break, so maybe worth waiting until the applications v2 release

It will be in some cases yes, but if we left it until the next major release then that just makes the problem more widespread. I think it's better to restore the correct behaviour for anyone new to using the gem. Existing gem users can upgrade through (recent changes since the bug was introduced are all backwards compatible API additions), and anyone wanting to stick with the buggy behaviour can lock the gem to the version they're currently using. Fixed in v5.7.1, sorry for the confusion.

**Comment by marklewin** [comment]
Hi @timcraft. Would you please review the attached snippets? I'm not a rubyist, but there doesn't seem to be any useful info in the `Nexmo::ClientError` to return.

**Comment by timcraft** [comment]
@marklewin If you want to keep the generic exception handling the best thing to do is rescue from `Nexmo::Error` and print out the exception message. For example:

```ruby
begin
  response = client.numbers.cancel(country: COUNTRY_CODE, msisdn: NEXMO_NUMBER)

  puts "Number cancelled"
rescue Nexmo::Error => exception
  puts "ERROR: #{exception.message}"
end
```

That'll catch any kind of response error. I'd argue it's better to leave that out of these snippets because exception handling varies a bit depending on the application/context, and running these snippets on the command line will show the exception message anyway so the code doesn't add much.

**Comment by marklewin** [comment]
Thanks @timcraft ! @mheap what are your thoughts on this?

**Comment by mheap** [comment]
I'd like to know more from @timcraft about how exception handling depends on context. Not showing how to catch exceptions is something we've had feedback about

**Comment by timcraft** [comment]
> I'd like to know more from @timcraft about how exception handling depends on context.

A few that come to mind: letting the exceptions bubble up to a generic exception handler and sending them to a monitoring service; letting them bubble up in a background job system which retries jobs automatically; catching and retrying them with exponential backoff. None of those require a begin..rescue block, only one needs the names of the exceptions.

> Not showing how to catch exceptions is something we've had feedback about

The only gem specific part of the exception catching is the name of the exception classes. I've started work on some better API docs for the gem which lists them, hopefully that will help.


The `client.numbers.list` and `client.numbers.search` calls can raise the same exceptions, so presumably better to be consistent and add rescue blocks for those too?

**Comment by marklewin** [comment]
Thanks @timcraft - I agree that consistency is good here, so have added exception handling to the `list.rb` and `search.rb` code snippets. Just out of interest: what would cause those operations to raise an exception?

**Comment by timcraft** [comment]
@marklewin Same as for any other operations:

HTTP 401 responses → `Nexmo::AuthenticationError` exceptions
HTTP 4xx responses → `Nexmo::ClientError` exceptions
HTTP 5xx responses → `Nexmo::ServerError` exceptions

**Comment by marklewin** [comment]
Thanks @timcraft . @mheap - could you approve this PR if you're happy with our approach please?

**Pull request merged**